File: src/TaskOTime/TaskOTime.ViewModel/ViewModels/CollaborationViewModel.vb
Imports System.Windows
Imports TaskOTime.AppServer.Models
Imports TaskOTime.ViewModel.Views

Namespace ViewModels
    Public Class CollaborationViewModel
        Private ReadOnly _view As CollaborationView
        Private ReadOnly _store As ServiceWorkspace
        Friend Sub New(view As CollaborationView, store As ServiceWorkspace)
            _view = view
            _store = store
            _view.DataContext = Me
       'tabs macht das viewmodel. so ist die UI logik zentraler und nicht überall
            _view.CategoryListView.ItemsSource = store.Categories
            _view.TagListBox.ItemsSource = store.Tags
            _view.NoteListView.ItemsSource = store.Notes
            _view.WebLinkListView.ItemsSource = store.WebLinks
            _view.LogListView.ItemsSource = store.Logs
            AddHandler _view.AddButton.Click, AddressOf AddCurrent
            AddHandler _view.DeleteButton.Click, AddressOf DeleteCurrent
        End Sub

        Private Sub AddCurrent(sender As Object, e As RoutedEventArgs)
            Dim value = _view.QuickValueTextBox.Text.Trim()
            If value.Length = 0 Then
                MessageBox.Show(Window.GetWindow(_view), "Bitte zuerst einen Wert eingeben.", "Hinzufügen")
                Return
            End If

            Select Case _view.DetailTabs.SelectedIndex
                Case 0
                    RunCreate("Kategorie anlegen", Sub()
                        Dim result = _store.AdminService.CreateCategory(New SaveCategoryRequest With {
                            .IdTenant = _store.Tenant.IdTenant, .IdActingUser = _store.ActingUserId,
                            .Item = New CategoryMasterDataDto With {
                                .IdTenant = _store.Tenant.IdTenant, .IdUser = _store.ActingUserId, .CategoryName = value
                            }
                        })
                        _store.Categories.Add(ServiceWorkspace.Require(result, "Kategorie anlegen"))
                    End Sub)
                Case 1
                    RunCreate("Tag anlegen", Sub()
                        Dim result = _store.AdminService.CreateTag(New SaveTagRequest With {
                            .IdTenant = _store.Tenant.IdTenant, .IdActingUser = _store.ActingUserId,
                            .Item = New TagMasterDataDto With {
                                .IdTenant = _store.Tenant.IdTenant, .IdUser = _store.ActingUserId,
                                .Tag = value, .DateCreated = DateTimeOffset.Now, .DateModified = DateTimeOffset.Now
                            }
                        })
                        _store.Tags.Add(ServiceWorkspace.Require(result, "Tag anlegen"))
                    End Sub)
                Case 2
                    RunCreate("Notiz anlegen", Sub()
                        Dim result = _store.AdminService.CreateNote(New SaveNoteRequest With {
                            .IdTenant = _store.Tenant.IdTenant, .IdActingUser = _store.ActingUserId,
                            .Item = New NoteMasterDataDto With {
                                .IdTenant = _store.Tenant.IdTenant, .IdUser = _store.ActingUserId,
                                .NoteText = value, .DateCreated = DateTimeOffset.Now, .DateModified = DateTimeOffset.Now
                            }
                        })
                        _store.Notes.Add(ServiceWorkspace.Require(result, "Notiz anlegen"))
                    End Sub)
                Case 3
                    RunCreate("Weblink anlegen", Sub()
                        Dim result = _store.AdminService.CreateWebLink(New SaveWebLinkRequest With {
                            .IdTenant = _store.Tenant.IdTenant, .IdActingUser = _store.ActingUserId,
                            .Item = New WebLinkMasterDataDto With {
                                .IdTenant = _store.Tenant.IdTenant, .IdUser = _store.ActingUserId,
                                .Title = value, .Link = value, .DateCreated = DateTimeOffset.Now, .DateModified = DateTimeOffset.Now
                            }
                        })
                        _store.WebLinks.Add(ServiceWorkspace.Require(result, "Weblink anlegen"))
                    End Sub)
                Case Else
                    _store.Logs.Add(value)
            End Select
            _view.QuickValueTextBox.Clear()
        End Sub

        Private Sub DeleteCurrent(sender As Object, e As RoutedEventArgs)
            Select Case _view.DetailTabs.SelectedIndex
                Case 0
                    Dim item = TryCast(_view.CategoryListView.SelectedItem, CategoryMasterDataDto)
                    If item IsNot Nothing Then RunDelete("Kategorie löschen",
                        _store.AdminService.DeleteCategory(DeleteRequest(item.IdCategory)),
                        Sub() _store.Categories.Remove(item))
                Case 1
                    Dim item = TryCast(_view.TagListBox.SelectedItem, TagMasterDataDto)
                    If item IsNot Nothing Then RunDelete("Tag löschen",
                        _store.AdminService.DeleteTag(DeleteRequest(item.IdTag)),
                        Sub() _store.Tags.Remove(item))
                Case 2
                    Dim item = TryCast(_view.NoteListView.SelectedItem, NoteMasterDataDto)
                    If item IsNot Nothing Then RunDelete("Notiz löschen",
                        _store.AdminService.DeleteNote(DeleteRequest(item.IdNote)),
                        Sub() _store.Notes.Remove(item))
                Case 3
                    Dim item = TryCast(_view.WebLinkListView.SelectedItem, WebLinkMasterDataDto)
                    If item IsNot Nothing Then RunDelete("Weblink löschen",
                        _store.AdminService.DeleteWebLink(DeleteRequest(item.IdWebLink)),
                        Sub() _store.WebLinks.Remove(item))
                Case Else
                    _store.Logs.Remove(TryCast(_view.LogListView.SelectedItem, String))
            End Select
        End Sub

        Private Function DeleteRequest(id As Guid) As DeleteMasterDataRequest
            Return New DeleteMasterDataRequest With {
                .IdTenant = _store.Tenant.IdTenant, .IdActingUser = _store.ActingUserId,
                .IdItem = id, .HardDelete = True
            }
        End Function

        Private Sub RunCreate(operation As String, action As Action)
            Try
                action()
            Catch ex As InvalidOperationException
                MessageBox.Show(Window.GetWindow(_view), ex.Message, operation & " - Servicefehler")
            End Try
        End Sub

        Private Sub RunDelete(operation As String, result As ServiceResult(Of MasterDataDeleteResult), remove As Action)
            Try
                ServiceWorkspace.Require(result, operation)
                remove()
            Catch ex As InvalidOperationException
                MessageBox.Show(Window.GetWindow(_view), ex.Message, operation & " - Servicefehler")
            End Try
        End Sub
    End Class
End Namespace

File: src/TaskOTime/TaskOTime.ViewModel/ViewModels/CollaborationViewModel.cs
using System;
using System.Collections.ObjectModel;
using TaskOTime.AppServer.Models;
using TaskOTime.ViewModel.Base;

namespace TaskOTime.ViewModel.ViewModels
{
    /// <summary>Creates and removes collaboration data using observable selection and commands.</summary>
    public sealed class CollaborationViewModel : MaintenanceViewModel
    {
        private int _selectedTab;
        private string _quickValue = "", _selectedLog;
        private CategoryMainDataDto _selectedCategory;
        private TagMainDataDto _selectedTag;
        private NoteMainDataDto _selectedNote;
        private WebLinkMainDataDto _selectedWebLink;

        public CollaborationViewModel(ServiceWorkspace store, IMaintenanceInteraction interaction) : base(store, interaction)
        {
            AddCommand = Command(AddCurrent, () => SelectedTab >= 0 && SelectedTab <= 4 && !string.IsNullOrWhiteSpace(QuickValue));
            DeleteCommand = Command(DeleteCurrent, HasSelection);
        }

        public ObservableCollection<CategoryMainDataDto> Categories => Store.Categories;
        public ObservableCollection<TagMainDataDto> Tags => Store.Tags;
        public ObservableCollection<NoteMainDataDto> Notes => Store.Notes;
        public ObservableCollection<WebLinkMainDataDto> WebLinks => Store.WebLinks;
        public ObservableCollection<string> Logs => Store.Logs;
        public DelegateCommand AddCommand { get; }
        public DelegateCommand DeleteCommand { get; }
        public int SelectedTab { get => _selectedTab; set { if (SetProperty(ref _selectedTab, value, nameof(SelectedTab))) RefreshCommands(); } }
        public string QuickValue { get => _quickValue; set { if (SetProperty(ref _quickValue, value, nameof(QuickValue))) RefreshCommands(); } }
        public CategoryMainDataDto SelectedCategory { get => _selectedCategory; set { if (SetProperty(ref _selectedCategory, value, nameof(SelectedCategory))) RefreshCommands(); } }
        public TagMainDataDto SelectedTag { get => _selectedTag; set { if (SetProperty(ref _selectedTag, value, nameof(SelectedTag))) RefreshCommands(); } }
        public NoteMainDataDto SelectedNote { get => _selectedNote; set { if (SetProperty(ref _selectedNote, value, nameof(SelectedNote))) RefreshCommands(); } }
        public WebLinkMainDataDto SelectedWebLink { get => _selectedWebLink; set { if (SetProperty(ref _selectedWebLink, value, nameof(SelectedWebLink))) RefreshCommands(); } }
        public string SelectedLog { get => _selectedLog; set { if (SetProperty(ref _selectedLog, value, nameof(SelectedLog))) RefreshCommands(); } }

        private bool HasSelection()
        {
            switch (SelectedTab)
            {
                case 0: return SelectedCategory != null;
                case 1: return SelectedTag != null;
                case 2: return SelectedNote != null;
                case 3: return SelectedWebLink != null;
                case 4: return SelectedLog != null;
                default: return false;
            }
        }

        private void AddCurrent()
        {
            var value = QuickValue.Trim();
            var tenant = Store.Tenant.IdTenant;
            var user = Store.ActingUserId;
            var now = DateTimeOffset.Now;
            switch (SelectedTab)
            {
                case 0:
                    var category = ServiceWorkspace.Require(Store.AdminService.CreateCategory(new SaveCategoryRequest
                    {
                        IdTenant = tenant, IdActingUser = user,
                        Item = new CategoryMainDataDto { IdTenant = tenant, IdUser = user, CategoryName = value }
                    }), "Create category");
                    Categories.Add(category);
                    SelectedCategory = category;
                    break;
                case 1:
                    var tag = ServiceWorkspace.Require(Store.AdminService.CreateTag(new SaveTagRequest
                    {
                        IdTenant = tenant, IdActingUser = user,
                        Item = new TagMainDataDto { IdTenant = tenant, IdUser = user, Tag = value, DateCreated = now, DateModified = now }
                    }), "Create tag");
                    Tags.Add(tag);
                    SelectedTag = tag;
                    break;
                case 2:
                    var note = ServiceWorkspace.Require(Store.AdminService.CreateNote(new SaveNoteRequest
                    {
                        IdTenant = tenant, IdActingUser = user,
                        Item = new NoteMainDataDto { IdTenant = tenant, IdUser = user, NoteText = value, DateCreated = now, DateModified = now }
                    }), "Create note");
                    Notes.Add(note);
                    SelectedNote = note;
                    break;
                case 3:
                    var webLink = ServiceWorkspace.Require(Store.AdminService.CreateWebLink(new SaveWebLinkRequest
                    {
                        IdTenant = tenant, IdActingUser = user,
                        Item = new WebLinkMainDataDto { IdTenant = tenant, IdUser = user, Title = value, Link = value, DateCreated = now, DateModified = now }
                    }), "Create web link");
                    WebLinks.Add(webLink);
                    SelectedWebLink = webLink;
                    break;
                case 4:
                    Logs.Add(value);
                    SelectedLog = value;
                    break;
                default: return;
            }
            QuickValue = "";
        }

        private void DeleteCurrent()
        {
            switch (SelectedTab)
            {
                case 0:
                    ServiceWorkspace.Require(Store.AdminService.DeleteCategory(DeleteRequest(SelectedCategory.IdCategory)), "Delete category");
                    Categories.Remove(SelectedCategory);
                    SelectedCategory = null;
                    break;
                case 1:
                    ServiceWorkspace.Require(Store.AdminService.DeleteTag(DeleteRequest(SelectedTag.IdTag)), "Delete tag");
                    Tags.Remove(SelectedTag);
                    SelectedTag = null;
                    break;
                case 2:
                    ServiceWorkspace.Require(Store.AdminService.DeleteNote(DeleteRequest(SelectedNote.IdNote)), "Delete note");
                    Notes.Remove(SelectedNote);
                    SelectedNote = null;
                    break;
                case 3:
                    ServiceWorkspace.Require(Store.AdminService.DeleteWebLink(DeleteRequest(SelectedWebLink.IdWebLink)), "Delete web link");
                    WebLinks.Remove(SelectedWebLink);
                    SelectedWebLink = null;
                    break;
                case 4:
                    Logs.Remove(SelectedLog);
                    SelectedLog = null;
                    break;
            }
        }

        private DeleteMainDataRequest DeleteRequest(Guid id) => new DeleteMainDataRequest
        {
            IdTenant = Store.Tenant.IdTenant, IdActingUser = Store.ActingUserId, IdItem = id, HardDelete = true
        };
    }
}
